server/src/main/java/org/elasticsearch/index/translog/TruncateTranslogCommand.java

```diff
@@ -135,6 +135,13 @@ protected void execute(Terminal terminal, OptionSet options, Environment env) th
                 commitData = commits.get(commits.size() - 1).getUserData();
                 String translogGeneration = commitData.get(Translog.TRANSLOG_GENERATION_KEY);
                 String translogUUID = commitData.get(Translog.TRANSLOG_UUID_KEY);
+                final long globalCheckpoint;
+                // In order to have a safe commit invariant, we have to assign max_seqno of the last commit to the global checkpoint.
```

Comment:
I think we need to do more here to get to a healthy shard - we need to also move the local checkpoint to the max seq # . Otherwise we have the weird state where local checkpoint < global checkpoint. By setting the local checkpoint to the max seq# we declare all missing ops (that were in the translog) as no ops.

Reply:
I would also like to double check why no test failed due to the above. We should get a tight grip on when our invariants (gcp <= lcp <= max seq#) kick in.